package/contents/ui/main.qml

```diff
@@ -96,6 +108,7 @@ PlasmaCore.Dialog {
     function hide() {
         focusTimer.running = false;
         mainDialog.visible = false;
+        mainDialog.tileShortcuts.clear();
```

Comment:
I haven't tested yet but since tile shortcuts are registered on `Component.onCompleted`, they only get registered once Exquisite starts along with Kwin. Wouldn't this make tile shortcuts unusable after hiding Exquisite once?

Reply:
Just saw your commit descriptions:
> We add the shortcuts to the list every time that we bring up the dialog,
which means that we slowly accumulate a decent number of them, and all
of the associated data structures.
>
> In the more extreme cases, this can absolutely end up with us taking out
kwin entirely with references to stale objects.

As I said above, the shortcuts are added on the `Component.onCompleted` callback of each WindowLayout button. `Component.onCompleted` runs only when the object is first instantiated, which is the start of KWin in Exquisite's case. Hiding and showing the Exquisite dialog doesn't re-instantiate objects.

Reply:
It didn't used to re-instantiate objects, but the change up at line 59 of main.qml changes that.

It's also, I believe, why I needed to do the complicated state dance on line 180 of WindowLayout.qml.

Without this change, but with the change to re-instantiate things every time we open the dialog, we end up continually adding to the list of shortcuts.

In the worst case, without some of the guard rails I added in, this results in kwin segfaulting, which wasn't something I expected to be possible with QML. :)